src/core/utils/multicall.ts

```diff
@@ -0,0 +1,108 @@
+import { BytesLike, providers } from 'ethers'
+import { Multicall } from '../../contracts/Multicall'
+import { logger } from './logger'
+
+interface PendingCall {
+  callParams: { target: string; callData: BytesLike; allowFailure: boolean }
+  resolve: (value: string) => void
+  reject: (reason: string | undefined) => void
+}
+
+export class MulticallQueue {
+  private pendingCalls: PendingCall[] = []
+  private multicallContract: Multicall
+  private flushTimeoutMs: number
+
+  private maxBatchSize: number
+  private _timer: NodeJS.Timeout | null = null
+
+  constructor({
+    flushTimeoutMs,
+    maxBatchSize,
+    multicallContract,
+  }: {
+    flushTimeoutMs: number
+    maxBatchSize: number
+    multicallContract: Multicall
+  }) {
+    this.flushTimeoutMs = flushTimeoutMs
+    this.maxBatchSize = maxBatchSize
+    this.multicallContract = multicallContract
+  }
+
+  private set timer(value: NodeJS.Timeout | null) {
+    if (this._timer) clearTimeout(this._timer)
+    this._timer = value
+  }
+
+  private get timer() {
+    return this._timer
+  }
+
+  async queueCall(callParams: providers.TransactionRequest): Promise<string> {
+    const { to, data } = callParams
+
+    if (!to || !data) {
+      throw new Error('To and Data are required when using multicall')
+    }
+
+    return new Promise((resolve, reject) => {
+      this.pendingCalls.push({
+        callParams: { target: to, callData: data, allowFailure: true },
+        resolve,
+        reject,
+      })
+
+      if (this.pendingCalls.length >= this.maxBatchSize) {
+        this._flush()
+      } else if (!this.timer) {
+        this.timer = setTimeout(() => {
+          this._flush()
+          this.timer = null
+        }, this.flushTimeoutMs)
+      }
+    })
+  }
+
+  private async _flush() {
+    const callsToProcess = [...this.pendingCalls]
+    this.pendingCalls = []
+
+    const batchSize = callsToProcess.length
+    logger.info({ batchSize }, 'Sending multicall batch ')
+
+    const results = await this.multicallContract.callStatic.aggregate3(
+      callsToProcess.map(({ callParams }) => callParams),
+    )
+
+    const resultLength = results.length
+    const callsToProcessLength = callsToProcess.length
+
+    if (resultLength !== callsToProcessLength) {
+      // reject all to be on safe side
+      callsToProcess.forEach(({ reject }) => {
+        reject('Multicall batch failed')
+      })
+
+      logger.error(
+        {
+          resultLength,
+          callsToProcessLength,
+        },
+        'Multicall response length differs from batch sent',
+      )
+      return
+    }
+
+    callsToProcess.forEach(({ resolve, reject }, i) => {
+      const returnedData = results[i]?.returnData
+      const success = results[i]?.success
```

Comment:
Regarding '?' I am not totally sure why it's not happy here. Might be something to do with the Multicall contract interface. I feel like I am covering issues with  `(resultLength !== callsToProcessLength)` check but happy to be proven wrong

Reply:
With `"noUncheckedIndexedAccess": true,`, any access via index will always return `T | undefined`.

If you know for sure from your internal implementation that it cannot be undefined because you have checked it beforehand, but TypeScript cannot parse it properly, then you need to use `results[i]!.success`.

Reply:
ok played alot with this and managed to get ts to stop complaining a simple if(!results[i]) didn't do the trick but assigning the value to a new var and checking that then stopped ts complaining 

```     
 const result = results[i]
      if (!result) return
```